Assignment 4/src/FractionalWrapper.sol

```diff
@@ -0,0 +1,318 @@
+    ///     underlying.
+    /// @dev Denoted in increments of 1e-27. `fractionDenominator` gives the
+    //      unit (corresponding to a fraction of 1:1).
+    ///
+    ///     Note: This conversion happens in terms of the smallest unit of the
+    ///     token. The wrapping token has a fixed number of decimals that may
+    ///     be different to the number of decimals of the underlying. If that
+    ///     is the case, an unintended conversion may happen. For example,
+    ///     wrapping a token with 0 decimals into a wrapper with 18 decimals
+    ///     will effectively correspond to a fraction of
+    ///     `1 / 1e-18 * fractionDenominator`.
+    uint256 public immutable fraction;
+    /// @notice The unit corresponding to a fraction of 1 in terms of the
+    ///     smallest token unit. See also `fraction`.
+    uint256 constant fractionDenominator = 10**27;
+
+    /// @notice The total amount of wrapped tokens in existence.
+    uint256 public override totalSupply = 0;
+
+    /// @notice The total amount of assets managed by the vault.
+    /// @dev This simply calls the balance of this address. This means that not
+    ///     all assets may be redeemable by an account: we could have received
+    ///     a generous gift. These would still add up to the total.
+    /// @return totalManagedAssets The assets managed by this contract.
+    function totalAssets() external override view returns (uint256 totalManagedAssets) {
+        return token.balanceOf(address(this));
+    }
+
+    /// @notice The current allowance by the owner to spend by the spender.
+    ///     Access as follows:
+    ///     - owner: The owner of the tokens. Note: The allowance may be higher
+    ///         than their current balance.
+    ///     - spender: The spender of the tokens.
+    mapping(address => mapping(address => uint256)) public override allowance;
+
+    /// @notice The balance of addresses in terms of wrapper tokens.
+    mapping(address => uint256) public override balanceOf;
+
+    /// @notice Create a new wrapper.
+    /// @param _token The token to use as collateral.
+    /// @param _fraction The fraction at which to mint/withdraw the tokens.
+    /// @dev See the caveats on `fraction` for how to use the `_fraction`
+    ///     parameter.
+    constructor(IERC20 _token, uint256 _fraction) {
+        token = _token;
+        fraction = _fraction;
+    }
+
+    /// @notice Transfer tokens to another account.
+    ///
+    ///     Fails with `TransferFailed` if the balance of the caller is not
+    ///     high enough.
+    /// @param recipient The tokens' new owner.
+    /// @param amount The amount of tokens to transfer. This is denominated in
+    ///     shares, not collateral!
+    /// @return successful whether the transfer was succesful. If the
+    ///     transaction did not revert, this always returns true.
+    function transfer(address recipient, uint256 amount) external override returns (bool) {
```

Comment:
Instead of typing each ERC20 function, inherit from yield-utils-v2 ERC20.sol in L17.